### PR TITLE
[SB-1093] Add support for converting v5 FindIntersect calls to v6 findIntersection

### PR DESCRIPTION
**Pull request by droark**
- Add code that can take a FindIntersect result and convert it to a findIntersection result.
- Update similar code as needed.

### CODE REVIEW

File: ouroboros/chainsync/types.go [hunk 1]
```diff
@@ -373,12 +552,235 @@ func (r Block) PointStruct() PointStruct {
...
+				References:               t.Body.References,
+				Collaterals:              t.Body.Collaterals,
+				TotalCollateral:          t.Body.TotalCollateral,
+				CollateralReturn:         (*TxOut)(t.Body.CollateralReturn),
+				Outputs:                  t.Body.Outputs,
+				Certificates:             t.Body.Certificates,
+				Withdrawals:              withdrawals,
+				Fee:                      Lovelace{Lovelace: t.Body.Fee, Extras: nil},
+				ValidityInterval:         t.Body.ValidityInterval,
+				Mint:                     nil, // TODO - Differences appear to be too much to handle.
```
Comment (Quantumplation) [inline ouroboros/chainsync/types.go:699]: This is one that we'll definitely need to handle, we depend on the mint field in plenty of cases

File: ouroboros/chainsync/types.go [hunk 2]
```diff
@@ -387,13 +789,268 @@ type Tip struct {
...
+		*c = CompatibleResponsePraos(r)
+		return nil
+	}
+
+	var r5 ResponseV5
+	err = json.Unmarshal(data, &r5)
+	c.JsonRpc = "2.0"
+	if err != nil {
+		// Just skip all the data processing, as it's useless.
+		return nil
```
Comment (Quantumplation) [inline ouroboros/chainsync/types.go:818]: We should return err instead of nil

File: ouroboros/chainsync/types.go [hunk 3]
```diff
@@ -387,13 +789,268 @@ type Tip struct {
...
+			t.ID = r5.Result.IntersectionNotFound.Tip.Hash
+			t.Height = r5.Result.IntersectionFound.Tip.BlockNo
+
+			var e ResultError
+			e.Data = &t
+			e.Code = 1000
+			e.Message = "Intersection not found - Conversion from a v5 Ogmigo call"
+			c.Error = &e
+		} else if r5.Result.RollForward != nil {
+			c.Method = NextBlockMethod
```
Comment (Quantumplation) [inline ouroboros/chainsync/types.go:858]: This looks like a lot of duplication with the above code; what's going on here?

File: ouroboros/chainsync/types.go (hunk 1 above)
Comment (droark) [reply ouroboros/chainsync/types.go]: So, `mint` is weird. I'm going to update the unit tests but here are examples of what we see in v5 and v6.

v5:
```
"mint": {
    "coins": 0,
    "assets": {
        "3542acb3a64d80c29302260d62c3b87a742ad14abf855ebc6733081e.00": 3,
        "e0a714319812c3f773ba04ec5d6b3ffcd5aad85006805b047b082541": 4,
        "e0a714319812c3f773ba04ec5d6b3ffcd5aad85006805b047b082541.00": -3
    }
}
```

v6:
```
"mint": {
    "be8dc1ca8b6b1735175a24b3b3c39d7327aba843c16a9b99924f3476": {
        "a3589dbc6194521a": 98371879850696154
    }
}
```

Looking at the API specs, it looks like v5 is `POLICY_ID.ASSET_NAME`, and v6 is just "property names." I'm stumped on this one.

File: ouroboros/chainsync/types.go (hunk 3 above)
Comment (droark) [reply ouroboros/chainsync/types.go]: Ahhh, I forgot to write some helper functions that cut down on duplication. I fixed that.

File: ouroboros/chainsync/v5/types.go [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+	// TODO: Further error handling here.
+	return nil
+}
+
+func (c CompatibleResultFindIntersection) String() string {
+	return fmt.Sprintf("intersection=[%v] tip=[%v] error=[%v] id=[%v]", c.Intersection, c.Tip, c.Error, c.ID)
+}
+
+// Support findIntersect (v6) / FindIntersection (v5) universally.
+type CompatibleResultNextBlock chainsync.ResultNextBlockPraos
```
Comment (Quantumplation) [inline ouroboros/chainsync/v5/types.go:413]: I would leave the CompatibleX types in the top level; `v5` should be entirely an implementation detail hidden to the external consumer (unless they absolutely need it).

Plus, doesn't this create an awkward circular dependency among the types? the parent module needs to depend on v5, but v5 now needs to depend on the parent module for `chainsync.ResultNextBlockPraos`

File: ouroboros/chainsync/types.go (hunk 1 above)
Comment (Quantumplation) [reply ouroboros/chainsync/types.go]: It looks like `chainsync.Value` has been updated from having a map from AssetId to amount (AssetId being `PolicyID.AssetNameHex`), to now being:
```
map[string]map[string]num.Int
```

i.e. a map from policy ID, to a map from asset name, to amount.

**Review by Quantumplation — APPROVED** [review]

File: ouroboros/chainsync/types.go (hunk 1 above)
Comment (droark) [reply ouroboros/chainsync/types.go]: Thanks. As I test `sundae-sync`, I'll fold in any relevant changes in the next PR.

**Pull request merged**